go/cmd/dolt/commands/backup.go

```diff
@@ -0,0 +1,337 @@
+	ap.SupportsString(dbfactory.AWSCredsProfile, "", "profile", "AWS profile to use")
+	return ap
+}
+
+// EventType returns the type of the event to log
+func (cmd BackupCmd) EventType() eventsapi.ClientEventType {
+	return eventsapi.ClientEventType_REMOTE
+}
+
+// Exec executes the command
+func (cmd BackupCmd) Exec(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
+	ap := cmd.createArgParser()
+	help, usage := cli.HelpAndUsagePrinters(cli.GetCommandDocumentation(commandStr, backupDocs, ap))
+	apr := cli.ParseArgsOrDie(ap, args, help)
+
+	var verr errhand.VerboseError
+
+	switch {
+	case apr.NArg() == 0:
+		verr = printBackups(dEnv, apr)
+	case apr.Arg(0) == addBackupId:
+		verr = addBackup(dEnv, apr)
+	case apr.Arg(0) == removeBackupId:
+		verr = removeBackup(ctx, dEnv, apr)
+	case apr.Arg(0) == removeBackupShortId:
+		verr = removeBackup(ctx, dEnv, apr)
+	case apr.Arg(0) == syncBackupId:
+		verr = syncBackup(ctx, dEnv, apr)
+	case apr.Arg(0) == restoreBackupId:
+		verr = restoreBackup(ctx, dEnv, apr)
+	default:
+		verr = errhand.BuildDError("").SetPrintUsage().Build()
+	}
+
+	return HandleVErrAndExitCode(verr, usage)
+}
+
+func removeBackup(ctx context.Context, dEnv *env.DoltEnv, apr *argparser.ArgParseResults) errhand.VerboseError {
+	if apr.NArg() != 2 {
+		return errhand.BuildDError("").SetPrintUsage().Build()
+	}
+
+	old := strings.TrimSpace(apr.Arg(1))
+	err := dEnv.RemoveBackup(ctx, old)
+
+	switch err {
+	case nil:
+		return nil
+	case env.ErrFailedToWriteRepoState:
+		return errhand.BuildDError("error: failed to save change to repo state").AddCause(err).Build()
+	case env.ErrFailedToDeleteBackup:
+		return errhand.BuildDError("error: failed to delete backup tracking ref").AddCause(err).Build()
+	case env.ErrFailedToReadFromDb:
+		return errhand.BuildDError("error: failed to read from db").AddCause(err).Build()
+	case env.ErrBackupNotFound:
+		return errhand.BuildDError("error: unknown backup: '%s' ", old).Build()
+	default:
+		return errhand.BuildDError("error: unknown error").AddCause(err).Build()
+	}
+}
+
+func addBackup(dEnv *env.DoltEnv, apr *argparser.ArgParseResults) errhand.VerboseError {
+	if apr.NArg() != 3 {
+		return errhand.BuildDError("").SetPrintUsage().Build()
+	}
+
+	backupName := strings.TrimSpace(apr.Arg(1))
+
+	backupUrl := apr.Arg(2)
+	scheme, absBackupUrl, err := env.GetAbsRemoteUrl(dEnv.FS, dEnv.Config, backupUrl)
+	if err != nil {
+		return errhand.BuildDError("error: '%s' is not valid.", backupUrl).AddCause(err).Build()
+	}
+
+	params, verr := parseBackupArgs(apr, scheme, absBackupUrl)
+	if verr != nil {
+		return verr
+	}
+
+	r := env.NewRemote(backupName, backupUrl, params, dEnv)
+	err = dEnv.AddBackup(r.Name, r.Url, r.FetchSpecs, r.Params)
+
+	switch err {
+	case nil:
+		return nil
+	case env.ErrBackupAlreadyExists:
+		return errhand.BuildDError("error: a backup named '%s' already exists.", r.Name).AddDetails("remove it before running this command again").Build()
```

Comment:
Seems like this will also trigger if there's a remote with the same name? Need a better error message.

Reply:
sorry, do you want it to error if there is a remote of the same name/url? they are different fields in the repo state, none of the AddBackup code intersects the remote list currently